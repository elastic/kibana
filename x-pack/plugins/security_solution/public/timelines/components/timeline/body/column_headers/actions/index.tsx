/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import { EuiButtonIcon } from '@elastic/eui';
import React from 'react';

import { ColumnHeaderOptions } from '../../../../../../timelines/store/timeline/model';
import { OnColumnRemoved } from '../../../events';
import { EventsHeadingExtra, EventsLoading } from '../../../styles';
import { Sort } from '../../sort';

import * as i18n from '../translations';

interface Props {
  header: ColumnHeaderOptions;
  isLoading: boolean;
  onColumnRemoved: OnColumnRemoved;
  sort: Sort;
}

/** Given a `header`, returns the `SortDirection` applicable to it */

export const CloseButton = React.memo<{
  columnId: string;
  onColumnRemoved: OnColumnRemoved;
}>(({ columnId, onColumnRemoved }) => (
  <EuiButtonIcon
    aria-label={i18n.REMOVE_COLUMN}
    color="text"
    data-test-subj="remove-column"
    iconType="cross"
    onClick={(event: React.MouseEvent<HTMLButtonElement>) => {
      // To avoid a re-sorting when you delete a column
      event.preventDefault();
      event.stopPropagation();
      onColumnRemoved(columnId);
    }}
  />
));

CloseButton.displayName = 'CloseButton';

export const Actions = React.memo<Props>(({ header, onColumnRemoved, sort, isLoading }) => {
  return (
    <>
      {sort.columnId === header.id && isLoading ? (
        <EventsHeadingExtra className="siemEventsHeading__extra--loading">
          <EventsLoading data-test-subj="timeline-loading-spinner" />
        </EventsHeadingExtra>
      ) : (
        <EventsHeadingExtra className="siemEventsHeading__extra--close">
          <CloseButton columnId={header.id} onColumnRemoved={onColumnRemoved} />
        </EventsHeadingExtra>
      )}
    </>
  );
});

Actions.displayName = 'Actions';
