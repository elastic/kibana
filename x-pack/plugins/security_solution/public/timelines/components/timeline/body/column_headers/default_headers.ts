/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { ColumnHeaderOptions, ColumnHeaderType } from '../../../../store/timeline/model';
import { DEFAULT_COLUMN_MIN_WIDTH, DEFAULT_DATE_COLUMN_MIN_WIDTH } from '../constants';

export const defaultColumnHeaderType: ColumnHeaderType = 'not-filtered';

export const defaultHeaders: ColumnHeaderOptions[] = [
  {
    columnHeaderType: defaultColumnHeaderType,
    id: '@timestamp',
    type: 'number',
    width: DEFAULT_DATE_COLUMN_MIN_WIDTH,
  },
  {
    columnHeaderType: defaultColumnHeaderType,
    id: 'message',
    width: DEFAULT_COLUMN_MIN_WIDTH,
  },
  {
    columnHeaderType: defaultColumnHeaderType,
    id: 'event.category',
    width: DEFAULT_COLUMN_MIN_WIDTH,
  },
  {
    columnHeaderType: defaultColumnHeaderType,
    id: 'event.action',
    width: DEFAULT_COLUMN_MIN_WIDTH,
  },
  {
    columnHeaderType: defaultColumnHeaderType,
    id: 'host.name',
    width: DEFAULT_COLUMN_MIN_WIDTH,
  },
  {
    columnHeaderType: defaultColumnHeaderType,
    id: 'source.ip',
    width: DEFAULT_COLUMN_MIN_WIDTH,
  },
  {
    columnHeaderType: defaultColumnHeaderType,
    id: 'destination.ip',
    width: DEFAULT_COLUMN_MIN_WIDTH,
  },
  {
    columnHeaderType: defaultColumnHeaderType,
    id: 'user.name',
    width: DEFAULT_COLUMN_MIN_WIDTH,
  },
];

/** The default category of fields shown in the Timeline */
export const DEFAULT_CATEGORY_NAME = 'default ECS';
