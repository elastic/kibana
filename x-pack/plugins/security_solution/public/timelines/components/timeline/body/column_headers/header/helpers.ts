/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { Direction } from '../../../../../../../common/search_strategy';
import { assertUnreachable } from '../../../../../../../common/utility_types';
import { ColumnHeaderOptions } from '../../../../../../timelines/store/timeline/model';
import { Sort, SortDirection } from '../../sort';

interface GetNewSortDirectionOnClickParams {
  clickedHeader: ColumnHeaderOptions;
  currentSort: Sort[];
}

/** Given a `header`, returns the `SortDirection` applicable to it */
export const getNewSortDirectionOnClick = ({
  clickedHeader,
  currentSort,
}: GetNewSortDirectionOnClickParams): Direction =>
  currentSort.reduce<Direction>(
    (acc, item) => (clickedHeader.id === item.columnId ? getNextSortDirection(item) : acc),
    Direction.desc
  );

/** Given a current sort direction, it returns the next sort direction */
export const getNextSortDirection = (currentSort: Sort): Direction => {
  switch (currentSort.sortDirection) {
    case Direction.desc:
      return Direction.asc;
    case Direction.asc:
      return Direction.desc;
    case 'none':
      return Direction.desc;
    default:
      return assertUnreachable(currentSort.sortDirection, 'Unhandled sort direction');
  }
};

interface GetSortDirectionParams {
  header: ColumnHeaderOptions;
  sort: Sort[];
}

export const getSortDirection = ({ header, sort }: GetSortDirectionParams): SortDirection =>
  sort.reduce<SortDirection>(
    (acc, item) => (header.id === item.columnId ? item.sortDirection : acc),
    'none'
  );

export const getSortIndex = ({ header, sort }: GetSortDirectionParams): number =>
  sort.findIndex((s) => s.columnId === header.id);
