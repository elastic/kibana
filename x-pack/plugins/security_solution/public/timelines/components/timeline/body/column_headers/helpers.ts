/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import { get } from 'lodash/fp';

import { BrowserFields } from '../../../../../common/containers/source';
import { ColumnHeaderOptions } from '../../../../../timelines/store/timeline/model';
import {
  DEFAULT_COLUMN_MIN_WIDTH,
  DEFAULT_DATE_COLUMN_MIN_WIDTH,
  SHOW_CHECK_BOXES_COLUMN_WIDTH,
  EVENTS_VIEWER_ACTIONS_COLUMN_WIDTH,
  DEFAULT_ACTIONS_COLUMN_WIDTH,
  MINIMUM_ACTIONS_COLUMN_WIDTH,
} from '../constants';

/** Enriches the column headers with field details from the specified browserFields */
export const getColumnHeaders = (
  headers: ColumnHeaderOptions[],
  browserFields: BrowserFields
): ColumnHeaderOptions[] => {
  return headers.map((header) => {
    const splitHeader = header.id.split('.'); // source.geo.city_name -> [source, geo, city_name]

    return {
      ...header,
      ...get(
        [splitHeader.length > 1 ? splitHeader[0] : 'base', 'fields', header.id],
        browserFields
      ),
    };
  });
};

export const getColumnWidthFromType = (type: string): number =>
  type !== 'date' ? DEFAULT_COLUMN_MIN_WIDTH : DEFAULT_DATE_COLUMN_MIN_WIDTH;

/** Returns the (fixed) width of the Actions column */
export const getActionsColumnWidth = (
  isEventViewer: boolean,
  showCheckboxes = false,
  additionalActionWidth = 0
): number => {
  const checkboxesWidth = showCheckboxes ? SHOW_CHECK_BOXES_COLUMN_WIDTH : 0;
  const actionsColumnWidth =
    checkboxesWidth +
    (isEventViewer ? EVENTS_VIEWER_ACTIONS_COLUMN_WIDTH : DEFAULT_ACTIONS_COLUMN_WIDTH) +
    additionalActionWidth;

  return actionsColumnWidth > MINIMUM_ACTIONS_COLUMN_WIDTH + checkboxesWidth
    ? actionsColumnWidth
    : MINIMUM_ACTIONS_COLUMN_WIDTH + checkboxesWidth;
};
