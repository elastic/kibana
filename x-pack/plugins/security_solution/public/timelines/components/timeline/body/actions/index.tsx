/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import React, { useCallback, useMemo } from 'react';
import { EuiButtonIcon, EuiCheckbox, EuiLoadingSpinner, EuiToolTip } from '@elastic/eui';
import {
  eventHasNotes,
  getEventType,
  getPinOnClick,
  InvestigateInResolverAction,
} from '../helpers';
import { AlertContextMenu } from '../../../../../detections/components/alerts_table/timeline_actions/alert_context_menu';
import { InvestigateInTimelineAction } from '../../../../../detections/components/alerts_table/timeline_actions/investigate_in_timeline_action';
import { AddEventNoteAction } from '../actions/add_note_icon_item';
import { PinEventAction } from '../actions/pin_event_action';
import { EventsTdContent } from '../../styles';
import * as i18n from '../translations';
import { DEFAULT_ICON_BUTTON_WIDTH } from '../../helpers';
import { useShallowEqualSelector } from '../../../../../common/hooks/use_selector';
import { AddToCaseAction } from '../../../../../cases/components/timeline_actions/add_to_case_action';
import { TimelineId, TimelineTabs } from '../../../../../../common/types/timeline';
import { timelineSelectors } from '../../../../store/timeline';
import { timelineDefaults } from '../../../../store/timeline/defaults';
import { Ecs } from '../../../../../../common/ecs';
import { inputsModel } from '../../../../../common/store';
import { TimelineNonEcsData } from '../../../../../../common/search_strategy/timeline';
import { OnPinEvent, OnRowSelected, OnUnPinEvent } from '../../events';
import { RowCellRender } from '../control_columns';

interface Props {
  ariaRowindex: number;
  action?: RowCellRender;
  width?: number;
  columnId: string;
  columnValues: string;
  checked: boolean;
  onRowSelected: OnRowSelected;
  eventId: string;
  loadingEventIds: Readonly<string[]>;
  onEventDetailsPanelOpened: () => void;
  showCheckboxes: boolean;
  data: TimelineNonEcsData[];
  ecsData: Ecs;
  index: number;
  eventIdToNoteIds: Readonly<Record<string, string[]>>;
  isEventPinned: boolean;
  isEventViewer?: boolean;
  onPinEvent: OnPinEvent;
  onUnPinEvent: OnUnPinEvent;
  refetch: inputsModel.Refetch;
  rowIndex: number;
  onRuleChange?: () => void;
  showNotes: boolean;
  tabType?: TimelineTabs;
  timelineId: string;
  toggleShowNotes: () => void;
}

export type ActionProps = Props;

const ActionsComponent: React.FC<ActionProps> = ({
  ariaRowindex,
  width,
  checked,
  columnValues,
  eventId,
  data,
  ecsData,
  eventIdToNoteIds,
  isEventPinned = false,
  isEventViewer = false,
  loadingEventIds,
  onEventDetailsPanelOpened,
  onPinEvent,
  onRowSelected,
  onUnPinEvent,
  refetch,
  onRuleChange,
  showCheckboxes,
  showNotes,
  timelineId,
  toggleShowNotes,
}) => {
  const emptyNotes: string[] = [];
  const getTimeline = useMemo(() => timelineSelectors.getTimelineByIdSelector(), []);

  const handleSelectEvent = useCallback(
    (event: React.ChangeEvent<HTMLInputElement>) =>
      onRowSelected({
        eventIds: [eventId],
        isSelected: event.currentTarget.checked,
      }),
    [eventId, onRowSelected]
  );
  const handlePinClicked = useCallback(
    () =>
      getPinOnClick({
        allowUnpinning: !eventHasNotes(eventIdToNoteIds[eventId]),
        eventId,
        onPinEvent,
        onUnPinEvent,
        isEventPinned,
      }),
    [eventIdToNoteIds, eventId, isEventPinned, onPinEvent, onUnPinEvent]
  );
  const timelineType = useShallowEqualSelector(
    (state) => (getTimeline(state, timelineId) ?? timelineDefaults).timelineType
  );
  const eventType = getEventType(ecsData);

  const isEventContextMenuEnabled = useMemo(
    () => !!ecsData.event?.kind && ecsData.event?.kind[0] === 'event',
    [ecsData.event?.kind]
  );

  return (
    <>
      {showCheckboxes && (
        <div key="select-event-container" data-test-subj="select-event-container">
          <EventsTdContent textAlign="center" width={DEFAULT_ICON_BUTTON_WIDTH}>
            {loadingEventIds.includes(eventId) ? (
              <EuiLoadingSpinner size="m" data-test-subj="event-loader" />
            ) : (
              <EuiCheckbox
                aria-label={i18n.CHECKBOX_FOR_ROW({ ariaRowindex, columnValues, checked })}
                data-test-subj="select-event"
                id={eventId}
                checked={checked}
                onChange={handleSelectEvent}
              />
            )}
          </EventsTdContent>
        </div>
      )}
      <div key="expand-event">
        <EventsTdContent textAlign="center" width={DEFAULT_ICON_BUTTON_WIDTH}>
          <EuiToolTip data-test-subj="expand-event-tool-tip" content={i18n.VIEW_DETAILS}>
            <EuiButtonIcon
              aria-label={i18n.VIEW_DETAILS_FOR_ROW({ ariaRowindex, columnValues })}
              data-test-subj="expand-event"
              iconType="arrowRight"
              onClick={onEventDetailsPanelOpened}
            />
          </EuiToolTip>
        </EventsTdContent>
      </div>
      <>
        <InvestigateInResolverAction
          ariaLabel={i18n.ACTION_INVESTIGATE_IN_RESOLVER_FOR_ROW({ ariaRowindex, columnValues })}
          key="investigate-in-resolver"
          timelineId={timelineId}
          ecsData={ecsData}
        />
        {timelineId !== TimelineId.active && eventType === 'signal' && (
          <InvestigateInTimelineAction
            ariaLabel={i18n.SEND_ALERT_TO_TIMELINE_FOR_ROW({ ariaRowindex, columnValues })}
            key="investigate-in-timeline"
            ecsRowData={ecsData}
            nonEcsRowData={data}
          />
        )}

        {!isEventViewer && (
          <>
            <AddEventNoteAction
              ariaLabel={i18n.ADD_NOTES_FOR_ROW({ ariaRowindex, columnValues })}
              key="add-event-note"
              showNotes={showNotes}
              toggleShowNotes={toggleShowNotes}
              timelineType={timelineType}
            />
            <PinEventAction
              ariaLabel={i18n.PIN_EVENT_FOR_ROW({ ariaRowindex, columnValues, isEventPinned })}
              key="pin-event"
              onPinClicked={handlePinClicked}
              noteIds={eventIdToNoteIds[eventId] || emptyNotes}
              eventIsPinned={isEventPinned}
              timelineType={timelineType}
            />
          </>
        )}
        {[
          TimelineId.detectionsPage,
          TimelineId.detectionsRulesDetailsPage,
          TimelineId.active,
        ].includes(timelineId as TimelineId) && (
          <AddToCaseAction
            ariaLabel={i18n.ATTACH_ALERT_TO_CASE_FOR_ROW({ ariaRowindex, columnValues })}
            key="attach-to-case"
            ecsRowData={ecsData}
          />
        )}
        <AlertContextMenu
          ariaLabel={i18n.MORE_ACTIONS_FOR_ROW({ ariaRowindex, columnValues })}
          key="alert-context-menu"
          ecsRowData={ecsData}
          timelineId={timelineId}
          disabled={eventType !== 'signal' && !isEventContextMenuEnabled}
          refetch={refetch}
          onRuleChange={onRuleChange}
        />
      </>
    </>
  );
};

ActionsComponent.displayName = 'ActionsComponent';

export const Actions = React.memo(ActionsComponent);
