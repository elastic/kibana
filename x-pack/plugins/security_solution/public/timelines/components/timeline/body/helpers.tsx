/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import React, { useCallback, useMemo } from 'react';
import { get, isEmpty } from 'lodash/fp';
import { useDispatch } from 'react-redux';

import { Ecs } from '../../../../../common/ecs';
import { TimelineItem, TimelineNonEcsData } from '../../../../../common/search_strategy';
import { updateTimelineGraphEventId } from '../../../store/timeline/actions';
import {
  TimelineEventsType,
  TimelineTypeLiteral,
  TimelineType,
} from '../../../../../common/types/timeline';
import { OnPinEvent, OnUnPinEvent } from '../events';
import { ActionIconItem } from './actions/action_icon_item';

import * as i18n from './translations';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const omitTypenameAndEmpty = (k: string, v: any): any | undefined =>
  k !== '__typename' && v != null ? v : undefined;

export const stringifyEvent = (ecs: Ecs): string => JSON.stringify(ecs, omitTypenameAndEmpty, 2);

export const eventHasNotes = (noteIds: string[]): boolean => !isEmpty(noteIds);

export const getPinTooltip = ({
  isPinned,
  // eslint-disable-next-line no-shadow
  eventHasNotes,
  timelineType,
}: {
  isPinned: boolean;
  eventHasNotes: boolean;
  timelineType: TimelineTypeLiteral;
}) =>
  timelineType === TimelineType.template
    ? i18n.DISABLE_PIN
    : isPinned && eventHasNotes
    ? i18n.PINNED_WITH_NOTES
    : isPinned
    ? i18n.PINNED
    : i18n.UNPINNED;

export interface IsPinnedParams {
  eventId: string;
  pinnedEventIds: Readonly<Record<string, boolean>>;
}

export const eventIsPinned = ({ eventId, pinnedEventIds }: IsPinnedParams): boolean =>
  pinnedEventIds[eventId] === true;

export interface GetPinOnClickParams {
  allowUnpinning: boolean;
  eventId: string;
  onPinEvent: OnPinEvent;
  onUnPinEvent: OnUnPinEvent;
  isEventPinned: boolean;
}

export const getPinOnClick = ({
  allowUnpinning,
  eventId,
  onPinEvent,
  onUnPinEvent,
  isEventPinned,
}: GetPinOnClickParams) => {
  if (!allowUnpinning) {
    return;
  }

  if (isEventPinned) {
    onUnPinEvent(eventId);
  } else {
    onPinEvent(eventId);
  }
};

/**
 * Creates mapping of eventID -> fieldData for given fieldsToKeep. Used to store additional field
 * data necessary for custom timeline actions in conjunction with selection state
 * @param timelineData
 * @param eventIds
 * @param fieldsToKeep
 */
export const getEventIdToDataMapping = (
  timelineData: TimelineItem[],
  eventIds: string[],
  fieldsToKeep: string[]
): Record<string, TimelineNonEcsData[]> =>
  timelineData.reduce((acc, v) => {
    const fvm = eventIds.includes(v._id)
      ? { [v._id]: v.data.filter((ti) => fieldsToKeep.includes(ti.field)) }
      : {};
    return {
      ...acc,
      ...fvm,
    };
  }, {});

export const isEventBuildingBlockType = (event: Ecs): boolean =>
  !isEmpty(event.signal?.rule?.building_block_type);

/** Return eventType raw or signal */
export const getEventType = (event: Ecs): Omit<TimelineEventsType, 'all'> => {
  if (!isEmpty(event.signal?.rule?.id)) {
    return 'signal';
  }
  return 'raw';
};

export const isInvestigateInResolverActionEnabled = (ecsData?: Ecs) =>
  get(['agent', 'type', 0], ecsData) === 'endpoint' &&
  get(['process', 'entity_id'], ecsData)?.length === 1 &&
  get(['process', 'entity_id', 0], ecsData) !== '';

interface InvestigateInResolverActionProps {
  timelineId: string;
  ecsData: Ecs;
}

const InvestigateInResolverActionComponent: React.FC<InvestigateInResolverActionProps> = ({
  timelineId,
  ecsData,
}) => {
  const dispatch = useDispatch();
  const isDisabled = useMemo(() => !isInvestigateInResolverActionEnabled(ecsData), [ecsData]);
  const handleClick = useCallback(
    () => dispatch(updateTimelineGraphEventId({ id: timelineId, graphEventId: ecsData._id })),
    [dispatch, ecsData._id, timelineId]
  );

  return (
    <ActionIconItem
      ariaLabel={i18n.ACTION_INVESTIGATE_IN_RESOLVER}
      content={i18n.ACTION_INVESTIGATE_IN_RESOLVER}
      dataTestSubj="investigate-in-resolver"
      iconType="node"
      id="investigateInResolver"
      isDisabled={isDisabled}
      onClick={handleClick}
    />
  );
};

InvestigateInResolverActionComponent.displayName = 'InvestigateInResolverActionComponent';

export const InvestigateInResolverAction = React.memo(InvestigateInResolverActionComponent);
