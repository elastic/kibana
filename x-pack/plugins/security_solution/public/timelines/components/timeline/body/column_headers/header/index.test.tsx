/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

import { mount, shallow } from 'enzyme';
import React from 'react';

import { timelineActions } from '../../../../../store/timeline';
import { TestProviders } from '../../../../../../common/mock';
import { ColumnHeaderType } from '../../../../../store/timeline/model';
import { Sort } from '../../sort';
import { CloseButton } from '../actions';
import { defaultHeaders } from '../default_headers';

import { HeaderComponent } from '.';
import { getNewSortDirectionOnClick, getNextSortDirection, getSortDirection } from './helpers';
import { Direction } from '../../../../../../../common/search_strategy';

const mockDispatch = jest.fn();
jest.mock('react-redux', () => {
  const original = jest.requireActual('react-redux');

  return {
    ...original,
    useSelector: jest.fn(),
    useDispatch: () => mockDispatch,
  };
});

const filteredColumnHeader: ColumnHeaderType = 'text-filter';

describe('Header', () => {
  const columnHeader = defaultHeaders[0];
  const sort: Sort[] = [
    {
      columnId: columnHeader.id,
      columnType: columnHeader.type ?? 'number',
      sortDirection: Direction.desc,
    },
  ];
  const timelineId = 'fakeId';

  test('renders correctly against snapshot', () => {
    const wrapper = shallow(
      <TestProviders>
        <HeaderComponent header={columnHeader} sort={sort} timelineId={timelineId} />
      </TestProviders>
    );
    expect(wrapper.find('HeaderComponent').dive()).toMatchSnapshot();
  });

  describe('rendering', () => {
    test('it renders the header text', () => {
      const wrapper = mount(
        <TestProviders>
          <HeaderComponent header={columnHeader} sort={sort} timelineId={timelineId} />
        </TestProviders>
      );

      expect(
        wrapper.find(`[data-test-subj="header-text-${columnHeader.id}"]`).first().text()
      ).toEqual(columnHeader.id);
    });

    test('it renders the header text alias when displayAsText is provided', () => {
      const displayAsText = 'Timestamp';
      const headerWithLabel = { ...columnHeader, displayAsText };
      const wrapper = mount(
        <TestProviders>
          <HeaderComponent header={headerWithLabel} sort={sort} timelineId={timelineId} />
        </TestProviders>
      );

      expect(
        wrapper.find(`[data-test-subj="header-text-${columnHeader.id}"]`).first().text()
      ).toEqual(displayAsText);
    });

    test('it renders the header as a `ReactNode` when `display` is provided', () => {
      const display: React.ReactNode = (
        <div data-test-subj="rendered-via-display">
          {'The display property renders the column heading as a ReactNode'}
        </div>
      );
      const headerWithLabel = { ...columnHeader, display };
      const wrapper = mount(
        <TestProviders>
          <HeaderComponent header={headerWithLabel} sort={sort} timelineId={timelineId} />
        </TestProviders>
      );

      expect(wrapper.find(`[data-test-subj="rendered-via-display"]`).exists()).toBe(true);
    });

    test('it prefers to render `display` instead of `displayAsText` when both are provided', () => {
      const displayAsText = 'this text should NOT be rendered';
      const display: React.ReactNode = (
        <div data-test-subj="rendered-via-display">{'this text is rendered via display'}</div>
      );
      const headerWithLabel = { ...columnHeader, display, displayAsText };
      const wrapper = mount(
        <TestProviders>
          <HeaderComponent header={headerWithLabel} sort={sort} timelineId={timelineId} />
        </TestProviders>
      );

      expect(wrapper.text()).toBe('this text is rendered via display');
    });

    test('it falls back to rendering header.id when `display` is not a valid React node', () => {
      const display = {}; // a plain object is NOT a `ReactNode`
      const headerWithLabel = { ...columnHeader, display };
      const wrapper = mount(
        <TestProviders>
          <HeaderComponent header={headerWithLabel} sort={sort} timelineId={timelineId} />
        </TestProviders>
      );

      expect(
        wrapper.find(`[data-test-subj="header-text-${columnHeader.id}"]`).first().text()
      ).toEqual(columnHeader.id);
    });

    test('it renders a sort indicator', () => {
      const headerSortable = { ...columnHeader, aggregatable: true };
      const wrapper = mount(
        <TestProviders>
          <HeaderComponent header={headerSortable} sort={sort} timelineId={timelineId} />
        </TestProviders>
      );

      expect(wrapper.find('[data-test-subj="header-sort-indicator"]').first().exists()).toEqual(
        true
      );
    });

    test('it renders a filter', () => {
      const columnWithFilter = {
        ...columnHeader,
        columnHeaderType: filteredColumnHeader,
      };

      const wrapper = mount(
        <TestProviders>
          <HeaderComponent header={columnWithFilter} sort={sort} timelineId={timelineId} />
        </TestProviders>
      );

      expect(wrapper.find('[data-test-subj="textFilter"]').first().props()).toHaveProperty(
        'placeholder'
      );
    });
  });

  describe('onColumnSorted', () => {
    test('it invokes the onColumnSorted callback when the header sort button is clicked', () => {
      const headerSortable = { ...columnHeader, aggregatable: true };
      const wrapper = mount(
        <TestProviders>
          <HeaderComponent header={headerSortable} sort={sort} timelineId={timelineId} />
        </TestProviders>
      );

      wrapper.find('[data-test-subj="header-sort-button"]').first().simulate('click');

      expect(mockDispatch).toBeCalledWith(
        timelineActions.updateSort({
          id: timelineId,
          sort: [
            {
              columnId: columnHeader.id,
              columnType: columnHeader.type ?? 'number',
              sortDirection: Direction.asc, // (because the previous state was Direction.desc)
            },
          ],
        })
      );
    });

    test('it does NOT render the header sort button when aggregatable is false', () => {
      const headerSortable = { ...columnHeader, aggregatable: false };
      const wrapper = mount(
        <TestProviders>
          <HeaderComponent header={headerSortable} sort={sort} timelineId={timelineId} />
        </TestProviders>
      );

      expect(wrapper.find('[data-test-subj="header-sort-button"]').length).toEqual(0);
    });

    test('it does NOT render the header sort button when aggregatable is missing', () => {
      const headerSortable = { ...columnHeader };
      const wrapper = mount(
        <TestProviders>
          <HeaderComponent header={headerSortable} sort={sort} timelineId={timelineId} />
        </TestProviders>
      );

      expect(wrapper.find('[data-test-subj="header-sort-button"]').length).toEqual(0);
    });

    test('it does NOT invoke the onColumnSorted callback when the header is clicked and aggregatable is undefined', () => {
      const mockOnColumnSorted = jest.fn();
      const headerSortable = { ...columnHeader, aggregatable: undefined };
      const wrapper = mount(
        <TestProviders>
          <HeaderComponent header={headerSortable} sort={sort} timelineId={timelineId} />
        </TestProviders>
      );

      wrapper.find(`[data-test-subj="header-${columnHeader.id}"]`).first().simulate('click');

      expect(mockOnColumnSorted).not.toHaveBeenCalled();
    });
  });

  describe('CloseButton', () => {
    test('it invokes the onColumnRemoved callback with the column ID when the close button is clicked', () => {
      const mockOnColumnRemoved = jest.fn();

      const wrapper = mount(
        <CloseButton columnId={columnHeader.id} onColumnRemoved={mockOnColumnRemoved} />
      );

      wrapper.find('[data-test-subj="remove-column"]').first().simulate('click');

      expect(mockOnColumnRemoved).toBeCalledWith(columnHeader.id);
    });
  });

  describe('getSortDirection', () => {
    test('it returns the sort direction when the header id matches the sort column id', () => {
      expect(getSortDirection({ header: columnHeader, sort })).toEqual(sort[0].sortDirection);
    });

    test('it returns "none" when sort direction when the header id does NOT match the sort column id', () => {
      const nonMatching: Sort[] = [
        {
          columnId: 'differentSocks',
          columnType: columnHeader.type ?? 'number',
          sortDirection: Direction.desc,
        },
      ];

      expect(getSortDirection({ header: columnHeader, sort: nonMatching })).toEqual('none');
    });
  });

  describe('getNextSortDirection', () => {
    test('it returns "asc" when the current direction is "desc"', () => {
      const sortDescending: Sort = {
        columnId: columnHeader.id,
        columnType: columnHeader.type ?? 'number',
        sortDirection: Direction.desc,
      };

      expect(getNextSortDirection(sortDescending)).toEqual('asc');
    });

    test('it returns "desc" when the current direction is "asc"', () => {
      const sortAscending: Sort = {
        columnId: columnHeader.id,
        columnType: columnHeader.type ?? 'number',
        sortDirection: Direction.asc,
      };

      expect(getNextSortDirection(sortAscending)).toEqual(Direction.desc);
    });

    test('it returns "desc" by default', () => {
      const sortNone: Sort = {
        columnId: columnHeader.id,
        columnType: columnHeader.type ?? 'number',
        sortDirection: 'none',
      };

      expect(getNextSortDirection(sortNone)).toEqual(Direction.desc);
    });
  });

  describe('getNewSortDirectionOnClick', () => {
    test('it returns the expected new sort direction when the header id matches the sort column id', () => {
      const sortMatches: Sort[] = [
        {
          columnId: columnHeader.id,
          columnType: columnHeader.type ?? 'number',
          sortDirection: Direction.desc,
        },
      ];

      expect(
        getNewSortDirectionOnClick({
          clickedHeader: columnHeader,
          currentSort: sortMatches,
        })
      ).toEqual(Direction.asc);
    });

    test('it returns the expected new sort direction when the header id does NOT match the sort column id', () => {
      const sortDoesNotMatch: Sort[] = [
        {
          columnId: 'someOtherColumn',
          columnType: columnHeader.type ?? 'number',
          sortDirection: 'none',
        },
      ];

      expect(
        getNewSortDirectionOnClick({
          clickedHeader: columnHeader,
          currentSort: sortDoesNotMatch,
        })
      ).toEqual(Direction.desc);
    });
  });

  describe('text truncation styling', () => {
    test('truncates the header text with an ellipsis', () => {
      const wrapper = mount(
        <TestProviders>
          <HeaderComponent header={columnHeader} sort={sort} timelineId={timelineId} />
        </TestProviders>
      );

      expect(
        wrapper.find(`[data-test-subj="header-text-${columnHeader.id}"]`).at(1)
      ).toHaveStyleRule('text-overflow', 'ellipsis');
    });
  });

  describe('header tooltip', () => {
    test('it has a tooltip to display the properties of the field', () => {
      const wrapper = mount(
        <TestProviders>
          <HeaderComponent header={columnHeader} sort={sort} timelineId={timelineId} />
        </TestProviders>
      );

      expect(wrapper.find('[data-test-subj="header-tooltip"]').exists()).toEqual(true);
    });
  });
});
