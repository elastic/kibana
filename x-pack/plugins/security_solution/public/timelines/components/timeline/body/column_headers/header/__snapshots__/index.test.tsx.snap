// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly against snapshot 1`] = `
<Fragment>
  <Memo(HeaderContentComponent)
    header={
      Object {
        "columnHeaderType": "not-filtered",
        "id": "@timestamp",
        "initialWidth": 190,
        "type": "number",
      }
    }
    isLoading={false}
    isResizing={false}
    onClick={[Function]}
    showSortingCapability={true}
    sort={
      Array [
        Object {
          "columnId": "@timestamp",
          "columnType": "number",
          "sortDirection": "desc",
        },
      ]
    }
  >
    <Actions
      header={
        Object {
          "columnHeaderType": "not-filtered",
          "id": "@timestamp",
          "initialWidth": 190,
          "type": "number",
        }
      }
      isLoading={false}
      onColumnRemoved={[Function]}
      sort={
        Array [
          Object {
            "columnId": "@timestamp",
            "columnType": "number",
            "sortDirection": "desc",
          },
        ]
      }
    />
  </Memo(HeaderContentComponent)>
  <Filter
    header={
      Object {
        "columnHeaderType": "not-filtered",
        "id": "@timestamp",
        "initialWidth": 190,
        "type": "number",
      }
    }
    onFilterChange={[Function]}
  />
</Fragment>
`;
